### PR TITLE
Changes to enable deployment to Heroku

### PR DESCRIPTION
**Pull request by AidenMontgomery**
To run the bot on Heroku I needed it to be compiled to JS.
I made some changes to the command and event loading to support loading .js files and making sure that the paths to the directories were correct.
The rest of the changes are around documenting the Heroku deployment process.

### CODE REVIEW

**Comment by thenoahhein** [comment]
Just want to confirm, we don't need the menuselection files anymore? Were we using them before with the dm-style approach?

**Comment by AidenMontgomery** [comment]
I removed the DM style slash command in the last PR and forgot to remove these files, so getting rid of them now.

File: Procfile [hunk 1]
```diff
@@ -0,0 +1 @@
+worker: node dist/index.js
```
Comment (crondinini) [inline Procfile:1]: Why did you go for worker instead of a normal web process? At least for now it seems that a web process would be enough, considering we don't take advantage of the possibilities of the worker.

**Review by crondinini — COMMENTED** [review]
Thanks for pushing this through! 🚀

File: Procfile (hunk 1 above)
Comment (AidenMontgomery) [reply Procfile:1]: I just assumed that a web worker was for HTTP hosted things and that the worker was for everything else.
It's also how I have my other bot I built deployed, with a website running on the web process and the bot running on the worker process.

Comment (crondinini) [reply Procfile:1]: Makes sense. I just wanted to make sure I wasn't missing something. The only difference from the code side is that you wouldn't need a Procfile with a web resource, since you could add the command to start. Really minor though, so I think it's fine to stay with the worker.

File: tsconfig.json [hunk 2]
```diff
@@ -16,7 +16,8 @@
   "inlineSourceMap": false,
   "inlineSources": true,
   "declaration": false,
+  "outDir": "./dist",
   // This avoids building the files.
-  "noEmit": true,
+  // "noEmit": true,
```
Comment (crondinini) [inline tsconfig.json]: Maybe worth removing this?

File: utils/classes.ts [hunk 3]
```diff
@@ -6,24 +6,29 @@ export class discordClient extends Client {
     [key: string]: any
 
     async loadCommandsToClient() {
-        const commandFiles = fs.readdirSync('././commands').filter(file => file.endsWith('.ts'));
+        console.log('Loading commands', `${__dirname}/../commands`);
```
Comment (crondinini) [inline utils/classes.ts]: Is this debug log still needed?

File: commands/addCategory.ts [hunk 4]
```diff
@@ -38,46 +38,55 @@ export async function execute(interaction: CommandInteraction) {
...
-    try {
-      const foundCategory = await findCategoryByName(category.trim());
-      if (foundCategory) {
-        await interaction.editReply('This category is already registered.');
-      }
-      else {
-        await createCategory(category.trim());
-        await interaction.editReply('Thank you. The category has been added.');
-      }
+  if (interactionMessage instanceof Message) {
```
Comment (crondinini) [inline commands/addCategory.ts]: Minor, but imo the code would be slightly more readable if we returned early instead of nesting the logic. So if `if(!(interactionMessage instanceof Message)) { return; }`

File: commands/addBlockchain.ts [hunk 5]
```diff
@@ -43,46 +43,55 @@ export async function execute(interaction: CommandInteraction) {
...
-    try {
-      const foundChain = await findBlockchainByName(blockchain.trim());
-      if (foundChain) {
-        await interaction.editReply('This blockchain is already registered.');
-      }
-      else {
-        await createBlockchain(blockchain.trim(), website ? website.trim() : website);
-        await interaction.editReply('Thank you. The blockchain has been added.');
-      }
+  if (interactionMessage instanceof Message) {
```
Comment (crondinini) [inline commands/addBlockchain.ts]: I also wondered what's the other types of interactionMessage and why we need this.

File: package.json [hunk 6]
```diff
@@ -1,9 +1,14 @@
...
+    "node": "16.6.0"
+  },
   "scripts": {
     "fmt": "eslint . --fix",
     "lint": "tsc --noEmit && eslint .",
     "deploy:commands": "ts-node deploy-commands.ts",
-    "dev": "ts-node index.ts"
+    "dev": "ts-node index.ts",
+    "build": "tsc",
+    "postinstall": "tsc"
```
Comment (crondinini) [inline package.json:11]: Why do we need both build and postinstall?

**Review by crondinini — APPROVED** [review]
Overall looks good to me. I just added a couple of minor suggestion comments.

File: tsconfig.json (hunk 2 above)
Comment (AidenMontgomery) [reply tsconfig.json]: Do you mean remove the entire line, because it is commented out (and the comment above), or remove the change, so that we set "noEmit": true?

File: commands/addCategory.ts (hunk 4 above)
Comment (AidenMontgomery) [reply commands/addCategory.ts]: I tried that initially and TypeScript still complained, let me try again, I might have missed a type check.
By that I mean, later in the code file it wouldn't let me use `interactionMessage` as a `Message`.

File: package.json (hunk 6 above)
Comment (AidenMontgomery) [reply package.json:11]: `build` is for local building, in case you want to check building and running JS on your own machine.
`postinstall` is something that Heroku will run after it does `npm install`.

File: commands/addBlockchain.ts (hunk 5 above)
Comment (AidenMontgomery) [reply commands/addBlockchain.ts]: `const interactionMessage: Message<true> | APIMessage | Message<boolean>`

File: commands/addCategory.ts (hunk 4 above)
Comment (AidenMontgomery) [reply commands/addCategory.ts]: OK, so maybe VS Code was acting up the first time I tried this. Now it looks like it is working!
I definitely prefer the early return, thanks for making me double check.

File: utils/classes.ts (hunk 3 above)
Comment (AidenMontgomery) [reply utils/classes.ts]: Nope, taking it out.

File: Procfile (hunk 1 above)
Comment (thenoahhein) [inline Procfile:1]: Perhaps we could add the deployment of commands on startup to the worker? That would make things less manual.

File: README.md [hunk 7]
```diff
@@ -61,8 +61,13 @@ Now, you can test out the slash commands you've created in the Discord server wh
 
 These are the current commands the bot supports:
 
+- `/add-author` - Add an author to the knowledge base
+- `/add-blockchain` - Add a blockchain to the knowledge base
+- `/add-category` - Add a category to the knowledge base
+- `/add-contributor` - Add a category to the knowledge base
+- `/add-resource` - Add a new resource to the knowledge base
+- `/add-tag` - Add a tag to the knowledge base
 - `/ping` - Send a test request to the bot
```
Comment (thenoahhein) [inline README.md]: petition to move /ping as it's just really there for testing. Yay or nay?

**Review by thenoahhein — APPROVED** [review]
Looks good to me, I just left a few comments, but nothing blocking.

File: README.md (hunk 7 above)
Comment (AidenMontgomery) [reply README.md]: Yeah, I think that's just in there from the guides followed when setting up discord.js
Happy to remove it from the bot completely

File: Procfile (hunk 1 above)
Comment (AidenMontgomery) [reply Procfile:1]: Yeah, I was thinking about that. I'll see if it is something that we can do as part of Heroku startup, but it might also be something we could put in a GitHub action or CI/CD pipeline if we get that far.

**Comment by AidenMontgomery** [comment]
I would recommend squashing this PR as it is merged, it got a bit messy with the bug fixes while testing in live!

**Comment by AidenMontgomery** [comment]
Everything has now been addressed, apart from maybe the Heroku deployment scripts in package.json
I think that we should have some more discussions in discord about how we move to a better CI/CD process for this project.
